src/core/inductive.ml

```diff
@@ -55,17 +61,67 @@ let gen_ind_typ_codom : popt -> sym -> (tbox list -> tbox) -> string -> tbox =
+  fun pos c ind_typ_list p_str x_str ->
+  let prop = _Symb c.symb_Prop in
+
+  let rec aux :
+    int -> (sym * (tvar * tbox)) list -> tbox list -> sym list
+    -> ((sym * (tvar * tbox)) list * tbox list) =
+    fun i assoc_predicates conclusion_list ind_typ_list ->
+    match ind_typ_list with
+    | [] -> assoc_predicates, conclusion_list
+    | ind_sym::q ->
+        (* STEP 1 - Create the predicate (name + type) associated to the
+           symbol [ind_sym]. *)
+          (* A - Create the name of the predicate. *)
+        let p_str = p_str ^ string_of_int i in
+        let p = Bindlib.new_var mkfree p_str in
+          (* B - Create the type of the predicate p. *)
+        let codom ts = _Impl (_Appl_symb ind_sym ts) prop in
+        let p_type = gen_ind_typ_codom pos ind_sym codom p_str in
+        (* STEP 2 - Create the conclusion of the inductive principle
+           associated to [ind_sym]. *)
+        let codom ts =
+          let x = Bindlib.new_var mkfree x_str in
+          let t = Bindlib.bind_var x (prf_of p c ts (_Vari x)) in
+          _Prod (_Appl_symb ind_sym ts) t
+        in
+        let conclusion = gen_ind_typ_codom pos ind_sym codom p_str in
+        aux (i+1) ((ind_sym, (p, p_type))::assoc_predicates)
+          (conclusion::conclusion_list) q
+  in
+  aux 0 [] [] ind_typ_list
+
 (** [fold_cons_typ pos codom inj_var build_rec_hyp domrec dom ind_sym cons_sym
-    f_rec f_not_rec init s] generates some value iteratively by going through
-    the structure of [cons_sym.sym_type]. The argument [pos] is the position
-    of the command inductive where the inductive type was defined. The symbol
-    [ind_sym] is the type of the current inductive type, and the symbol
-    [cons_sym] is the current constructor. If you would like to store a
-    temporary result, the initial value is [init], and you can change this
-    value in the recursive case with the function [f_rec res x rec_hyp], and
-    on the other case with the function [f_not_rec res x]. The string [s] is
-    the prefix of variables' name. It's useful for the function [inj_var] to
-    have names with no clash.
+    f_rec f_not_rec init s assoc_predicates] generates some value iteratively
+    by going through the structure of [cons_sym.sym_type]. The argument [pos]
+    is the position of the command inductive where the inductive type was
+    defined. The symbol [ind_sym] is the type of the current inductive type,
+    and the symbol [cons_sym] is the current constructor. If you would like to
+    store a temporary result, the initial value is [init], and you can change
+    this value in the recursive case with the function [f_rec res x rec_hyp],
+    and on the other case with the function [f_not_rec res x]. The string [s]
+    is the prefix of variables' name. It's useful for the function [inj_var]
+    to have names with no clash. The data structure [assoc_predicates] stores
+    the link between an inductive type and a predicate with its type.
```

Comment:
[assoc_predicate] maps every inductive type to a predicate variable and its type.